crates/core_simd/src/swizzle.rs

```diff
@@ -0,0 +1,297 @@
+use crate::simd::intrinsics;
+use crate::{LaneCount, Simd, SimdElement, SupportedLaneCount};
+
+/// An index into one of two vectors.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum Which {
+    /// Indexes the first vector.
+    First(usize),
+    /// Indexes the second vector.
+    Second(usize),
+}
+
+/// Create a vector from the elements of another vector.
+pub trait Swizzle<const INPUT_LANES: usize, const OUTPUT_LANES: usize> {
+    /// Map from the lanes of the input vector to the output vector.
+    const INDEX: [usize; OUTPUT_LANES];
+
+    #[doc(hidden)]
+    /// The `simd_shuffle` intrinsic expects `u32`, so do error checking and conversion here.
+    const INDEX_IMPL: [u32; OUTPUT_LANES] = check::<INPUT_LANES, OUTPUT_LANES>(Self::INDEX);
```

Comment:
this should be somehow protected, since a user can currently cause UB without `unsafe` by overriding `INDEX_IMPL`.

Reply:
Actually, it wouldn't cause UB, the compiler would bail with a monomorphisation error if we'rw ok with that.

Reply:
we could do something like:
```rust
mod private {
pub trait SwizzleConsts<const A: usize, const B: usize>: Swizzle<A, B> {
    const INDEX_IMPL: [...; ...];
}
}

pub trait Swizzle<const A: usize, const B: usize> {
    const INDEX: [...; ...];
    fn swizzle(self) -> ...
    where
        Self: private::SwizzleConsts<A, B>,
    {
        ...
    }
}

impl<T, const A: usize, const B: usize> private::SwizzleConsts<A, B> for T
where
    T: Swizzle<A, B>,
{
    ...
}
```

Reply:
Inside the stdlib this stuff often can be protected using `#[unstable]`, even if its a bit of a hack.

Reply:
I was able to hide the impls with a helper trait like this